lib/explorer/backend/data_frame.ex

```diff
@@ -231,56 +231,63 @@ defmodule Explorer.Backend.DataFrame do
   end
 
   @default_limit 5
-  import Inspect.Algebra
+  alias Inspect.Algebra, as: A
 
   @doc """
   Default inspect implementation for backends.
   """
   def inspect(df, backend, n_rows, inspect_opts, opts \\ [])
       when is_binary(backend) and (is_integer(n_rows) or is_nil(n_rows)) and is_list(opts) do
     inspect_opts = %{inspect_opts | limit: @default_limit}
-    open = color("[", :list, inspect_opts)
-    close = color("]", :list, inspect_opts)
+    open = A.color("[", :list, inspect_opts)
+    close = A.color("]", :list, inspect_opts)
 
     cols_algebra =
       for name <- DataFrame.names(df) do
         series = df[name]
 
-        values =
+        data =
           series
           |> Series.slice(0, inspect_opts.limit + 1)
           |> Series.to_list()
-
-        data = container_doc(open, values, close, inspect_opts, &Explorer.Shared.to_string/2)
+          |> to_doc(inspect_opts)
 
         type =
           series
           |> Series.dtype()
           |> Explorer.Shared.dtype_to_string()
 
-        concat([
-          line(),
-          color("#{name} ", :map, inspect_opts),
-          color("#{type} ", :atom, inspect_opts),
+        A.concat([
+          A.line(),
+          A.color("#{name} ", :map, inspect_opts),
+          A.color("#{type} ", :atom, inspect_opts),
           data
         ])
       end
 
-    concat([
-      color(backend, :atom, inspect_opts),
+    A.concat([
+      A.color(backend, :atom, inspect_opts),
       open,
       "#{n_rows || "???"} x #{length(cols_algebra)}",
       close,
       groups_algebra(df.groups, inspect_opts) | cols_algebra
     ])
   end
 
+  defp to_doc(item, opts) when is_list(item) do
+    open = A.color("[", :list, opts)
+    close = A.color("]", :list, opts)
+    A.container_doc(open, item, close, opts, &to_doc/2)
+  end
+
+  defp to_doc(item, _opts), do: Explorer.Shared.to_string(item)
```

Comment:
Oh, it seems we only call `Shared.to_string` on inspect. So my suggestion is to move `to_doc` to `Explorer.Shared.to_doc` and merge the to_string implementation into it.

Reply:
Done! 504531c5fd1b75074d4d1560738dbe3ce231c414